src/Cache/FileCache.php

```diff
@@ -129,6 +129,15 @@ public function clear($cid = null, $wildcard = false)
         $bin_dir = $this->cacheDirectory();
         $files = [];
         if (empty($cid)) {
+            // TODO: Extract this whole bit into a generic "remove" function so it can be reused below for specific bins.
+            $semaphore = Path::join($bin_dir, 'semaphore');
```

Comment:
I don't think it works to put your semaphore in `$bin_dir`. You want the semaphore to be deleted at the end of the cache clear, and you cannot control what order the contents of `$bin_dir` will be deleted. The semaphore might even be deleted first.

You might consider using [flock](http://php.net/manual/en/function.flock.php) on a file parallel to $bin_dir (e.g. `$bin_dir.lock`)

Reply:
Makes sense, although for what it's worth this does seem to work fine in practice. I think Symfony Filesystem takes a depth-first approach to deleting directories, so the semaphore file in the root-level cache directory is always one of the last to get removed.